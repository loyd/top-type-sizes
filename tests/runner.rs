use structopt::StructOpt;

use top_type_sizes::*;

fn test(content: &'static str) {
    let refined_content = reader::read(content.as_bytes()).unwrap();

    let types = parser::parse(&refined_content).unwrap();
    insta::with_settings!({ description => "internal representation" }, {
        insta::assert_yaml_snapshot!("types", types);
    });

    snap_output(&types, &[]);
    snap_output(&types, &["-ws"]);
    snap_output(&types, &["-h8"]);
    snap_output(&types, &["-ws", "-h16"]);
}

fn snap_output(types: &[schema::Type], cmd: &[&str]) {
    let full_cmd = std::iter::once(&"top-type-sizes")
        .chain(cmd)
        .copied()
        .collect::<Vec<_>>();

    let options = options::Options::from_iter(&full_cmd);
    let snap_name = {
        let mut parts = cmd.to_vec();
        parts.insert(0, "output");
        parts.join("")
    };

    let types = transformer::transform(types.to_vec(), &options);
    let output = formatter::format(types, &options);

    insta::with_settings!({ description => full_cmd.join(" ") }, {
        insta::assert_snapshot!(snap_name, output);
    })
}

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            insta::with_settings!(
            {
                snapshot_path => concat!("snapshots/", stringify!($name)),
                prepend_module_to_snapshot => false,
                omit_expression => true,
            },
            {
                test(include_str!(concat!("samples/", stringify!($name), ".txt")))
            })
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
