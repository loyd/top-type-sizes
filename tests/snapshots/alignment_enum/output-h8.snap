---
source: tests/runner.rs
description: top-type-sizes -h8
---
8 std::ptr::alignment::AlignmentEnum64 align=8
      8 <discriminant>
      0 variant _Align1Shl0, _Align1Shl1, _Align1Shl2, _Align1Shl3, _Align1Shl4, _Align1Shl5, _Align1Shl6, _Align1Shl7, _Align1Shl8, _Align1Shl9, _Align1Shl10, _Align1Shl11, _Align1Shl12, _Align1Shl13, _Align1Shl14, _Align1Shl15, _Align1Shl16, _Align1Shl17, _Align1Shl18, _Align1Shl19, _Align1Shl20, _Align1Shl21, _Align1Shl22, _Align1Shl23, _Align1Shl24, _Align1Shl25, _Align1Shl26, _Align1Shl27, _Align1Shl28, _Align1Shl29, _Align1Shl30, _Align1Shl31, _Align1Shl32, _Align1Shl33, _Align1Shl34, _Align1Shl35, _Align1Shl36, _Align1Shl37, _Align1Shl38, _Align1Shl39, _Align1Shl40, _Align1Shl41, _Align1Shl42, _Align1Shl43, _Align1Shl44, _Align1Shl45, _Align1Shl46, _Align1Shl47, _Align1Shl48, _Align1Shl49, _Align1Shl50, _Align1Shl51, _Align1Shl52, _Align1Shl53, _Align1Shl54, _Align1Shl55, _Align1Shl56, _Align1Shl57, _Align1Shl58, _Align1Shl59, _Align1Shl60, _Align1Shl61, _Align1Shl62, _Align1Shl63
